### PR TITLE
Adjust error tolerance for oneDNN nondeterministic

### PR DESCRIPTION
**Pull request by Kanya-Mo**
This is to fix https://github.com/intel/torch-xpu-ops/issues/1682

Some of the distributed UTs are failed due to accuracy gap introduced by oneDNN non-deterministic and this gap is not distributed specific. Following previous practice, this PR increases their error tolerance.

`precisionOverride` cannot control behavior of `torch.testing.assert_close`, so I replaced them with `self.assertEqual` since these two are equivalent in terms of comparing. If this is not acceptable for upstreaming, I can instead explicitly adjust atol/rtol in `torch.testing.assert_close`.

### CODE REVIEW

File: test/distributed/pipelining/test_backward.py [hunk 1]
```diff
@@ -19,7 +23,9 @@
 
 
 class StageBackwardTests(TestCase):
-    def test_stage_backward(self, device):
+    @precisionOverride({torch.float: 1e-4})
```
Comment (daisyden) [inline test/distributed/pipelining/test_backward.py]: We will need to upstream this branch in next step, so the test case should work for all the backends. We should not change the test scope for cuda and other backends.

Comment (Kanya-Mo) [reply test/distributed/pipelining/test_backward.py]: Updated with a condition: precision override only happens when testing on xpu device. Do you think we should consider the case where xpu and other backends testing together on the same machine? If not, this condition should be fine.

File: test/distributed/pipelining/test_backward.py [hunk 2]
```diff
@@ -10,16 +10,22 @@
...
 
 
 d_hid = 512
 batch_size = 256
 
 
 class StageBackwardTests(TestCase):
-    def test_stage_backward(self, device):
+    @precisionOverride({torch.float: 1e-4} if TEST_XPU else {})
+    @dtypes(torch.float)
```
Comment (daisyden) [inline test/distributed/pipelining/test_backward.py]: What is the purpose to add @dtypes? Seems dtype is not not used in the case.

**Comment by daisyden** [comment]
The logic of torch.testing.assert_close() is not obviously the same as self.assertEqual. I suggest not to use @precisionOverride, instead use the torch.testing.assert_close() augments:
```
rtol: Optional[float] = None,
atol: Optional[float] = None,
```
For example:
```
if TEST_XPU:
     torch.testing.assert_close(p.grad, ref_p.grad, rtol=1e-4,  atol=1e-4)
else:
     torch.testing.assert_close(p.grad, ref_p.grad)
```

File: test/distributed/pipelining/test_backward.py (hunk 2 above)
Comment (Kanya-Mo) [reply test/distributed/pipelining/test_backward.py]: @precisionOverride only works with UTs that instantiated with dtype.

**Comment by Kanya-Mo** [comment]
> The logic of torch.testing.assert_close() is not obviously the same as self.assertEqual. I suggest not to use @precisionOverride, instead use the torch.testing.assert_close() augments:
> 
> ```
> rtol: Optional[float] = None,
> atol: Optional[float] = None,
> ```
> 
> For example:
> 
> ```
> if TEST_XPU:
>      torch.testing.assert_close(p.grad, ref_p.grad, rtol=1e-4,  atol=1e-4)
> else:
>      torch.testing.assert_close(p.grad, ref_p.grad)
> ```

Sounds good. In terms of comparing, those two APIs wrap the same thing `not_close_error_metas`, but yes, I agree that use `torch.testing.assert_close` augments is a cleaner solution here.

File: test/distributed/pipelining/test_backward.py [hunk 3]
```diff
@@ -124,11 +128,15 @@ def test_stage_backward_weight(self, device):
         ref_loss = loss_fn(ref_out, ref_target)
         ref_loss.backward()
 
+        rtol, atol = None, None
+        if self.device_type == "xpu":
+            rtol, atol = torch.testing._comparison.default_tolerances(torch.float32)[0], 1e-4
```
Comment (daisyden) [inline test/distributed/pipelining/test_backward.py:133]: If rtol is going to use default, why not just use 'None'?

Comment (Kanya-Mo) [reply test/distributed/pipelining/test_backward.py:133]: If `atol` is specified, `rtol` must also be specified. If set only `rtol` to None, it will raise ValueError.

**Review by daisyden — APPROVED** [review]
LGTM

**Comment by guangyey** [comment]
Hi @Kanya-Mo 
I see `MLPModule` is very straightforward. I don't understand how non-deterministic will impact the accuracy. Could you help elaborate on the accuracy gap.
```python
class MLPModule(torch.nn.Module):
    def __init__(self, d_hid: int):
        super().__init__()
        self.net1 = torch.nn.Linear(d_hid, d_hid)
        self.relu = torch.nn.ReLU()
        self.net2 = torch.nn.Linear(d_hid, d_hid)

    def forward(self, x):
        x = self.net1(x)
        x = self.relu(x)
        x = self.net2(x)
        return x
```

**Comment by Kanya-Mo** [comment]
> Hi @Kanya-Mo I see `MLPModule` is very straightforward. I don't understand how non-deterministic will impact the accuracy. Could you help elaborate on the accuracy gap.
> 
> ```python
> class MLPModule(torch.nn.Module):
>     def __init__(self, d_hid: int):
>         super().__init__()
>         self.net1 = torch.nn.Linear(d_hid, d_hid)
>         self.relu = torch.nn.ReLU()
>         self.net2 = torch.nn.Linear(d_hid, d_hid)
> 
>     def forward(self, x):
>         x = self.net1(x)
>         x = self.relu(x)
>         x = self.net2(x)
>         return x
> ```

Hi @guangyey, these distributed UTs compare two copies of a same model, with one executes in a distributed way (to test certain features) and the other runs in a straightforward way, and require the difference between their output under a tolerance of 1e-5. After some investigation I found on single device or even after separating out distributed features, these UTs still fail. So essentially, the difference from run to run itself will exceed the tolerance. 

Some ops are executed in a nondeterministic way (different orders of accumulation between runs). While the difference is very small for a single op, in a model (as simple as MLP) these differences could accumulate and fail these UTs. The difference will disappear if setting onednn deterministic flag.

I think this is known behavior and there are probably similar cases in single device UTs like [this one](https://github.com/pytorch/pytorch/blob/main/test/xpu/test_gemm.py#L233).

**Comment by guangyey** [comment]
@Kanya-Mo Thanks for your elaboration; `torch.nn.Linear` will invoke oneDNN gemm, so it makes sense.
